File: src/main/java/com/lessmarkup/engine/data/DomainModelProviderImpl.java
package com.lessmarkup.engine.data;

import com.google.inject.Inject;
import com.lessmarkup.interfaces.data.DataObject;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.module.ModuleConfiguration;
import com.lessmarkup.interfaces.module.ModuleProvider;

import java.util.OptionalInt;

class DomainModelProviderImpl implements DomainModelProvider {

    private final ModuleProvider moduleProvider;

    @Inject
    public DomainModelProviderImpl(ModuleProvider moduleProvider) {
        this.moduleProvider = moduleProvider;
    }
    
    @Override
    public void initialize() {
        for (ModuleConfiguration module : moduleProvider.getModules()) {
            for (Class<? extends DataObject> dataObject : module.getInitializer().getDataObjectTypes()) {
                DomainModelImpl.registerDataType(dataObject);
            }
        }
    }
    
    @Override
    public DomainModel create() {
        DomainModelImpl domainModel = new DomainModelImpl();
        domainModel.createConnection(null);
        return domainModel;
    }

    @Override
    public DomainModel create(String connectionString) {
        DomainModelImpl domainModel = new DomainModelImpl();
        domainModel.createConnection(connectionString);
        return domainModel;
    }

    @Override
    public DomainModel createWithTransaction() {
        DomainModelImpl domainModel = new DomainModelImpl();
        domainModel.createConnectionWithTransaction();
        return domainModel;
    }

    @Override
    public OptionalInt getCollectionId(Class<?> collectionType) {
        return DomainModelImpl.getCollectionId(collectionType);
    }
}

File: src/main/scala/com/lessmarkup/engine/data/DomainModelProviderImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.data

import com.google.inject.Inject
import com.lessmarkup.interfaces.annotations.Implements
import com.lessmarkup.interfaces.data.{DomainModel, DomainModelProvider}
import com.lessmarkup.interfaces.module.ModuleProvider

@Implements(classOf[DomainModelProvider])
class DomainModelProviderImpl @Inject() (moduleProvider: ModuleProvider) extends DomainModelProvider {

  def initialize() {
    for (
      module <- moduleProvider.getModules;
      dataObject <- module.getInitializer.getDataObjectTypes
    ) {
        MetadataStorage.registerDataType(dataObject)
    }
  }

  def create: DomainModel = {
    new DomainModelImpl(scala.Option.empty, false)
  }

  def create(connectionString: String): DomainModel = {
    new DomainModelImpl(scala.Option.apply(connectionString), false)
  }

  def createWithTransaction: DomainModel = {
    new DomainModelImpl(scala.Option.empty, true)
  }

  def getCollectionId(collectionType: Class[_]): Option[Int] = {
    MetadataStorage.getCollectionId(collectionType)
  }
}
